File: libraries/stzlib/base/test/temp.pl


:- use_module(library(http/json)).

% Main function to transform Prolog term to Ring format and save to file
transform_to_ring(Term, Filename) :-
    transform(Term, ResultStr),
    open(Filename, write, Stream),
    write(Stream, ResultStr),
    close(Stream).

% Transform an atom or string
transform(Term, Result) :-
    (atom(Term) ; string(Term)),
    !,
    format(atom(Result), "\'~w\'", [Term]).

% Transform a number, handling scientific notation
transform(Term, Result) :-
    number(Term),
    !,
    format(atom(StrVal), "~w", [Term]),
    (   sub_atom(StrVal, _, _, _, e) 
    ->  format(atom(Result), "\'~w\'", [Term])
    ;   Result = StrVal
    ).

% Transform boolean true value
transform(true, "TRUE") :- !.

% Transform boolean false value
transform(false, "FALSE") :- !.

% Transform a list
transform(List, Result) :-
    is_list(List),
    !,
    transform_list(List, Result).

% Transform a compound term
transform(Term, Result) :-
    compound(Term),
    !,
    Term =.. [Functor|Args],
    transform_compound(Functor, Args, Result).

% Default case for variables or other terms
transform(_, "NULL").

% Transform a list into a flat string
transform_list(List, Result) :-
    transform_list_items(List, Items),
    format(atom(Result), "[~w]", [Items]).

% Transform list items into a comma-separated string
transform_list_items([], "").
transform_list_items([H], Result) :-
    transform(H, HResult),
    format(atom(Result), "~w", [HResult]).
transform_list_items([H|T], Result) :-
    transform(H, HResult),
    transform_list_items(T, TResult),
    (   TResult = "" -> format(atom(Result), "~w", [HResult])
    ;   format(atom(Result), "~w, ~w", [HResult, TResult])
    ).

% Handle key-value pairs (Term = Key-Value), including numeric keys
transform_compound(-, [Key, Value], Result) :-
    transform(Key, KeyResult),
    transform(Value, ValueResult),
    format(atom(Result), "[~w, ~w]", [KeyResult, ValueResult]).

% Default handling for other compound terms
transform_compound(_, Args, Result) :-
    transform_list(Args, Result).


:- use_module(library(lists)).
:- use_module(library(apply)).

% User predicates


:- use_module(library(random)).
:- use_module(library(clpfd)).

% Fibonacci calculation (iterative using accumulator)

fib(N, Result) :-
    N =< 1,
    Result #= N.
fib(N, Result) :-
    N > 1,
    fib_iter(N, 0, 1, Result).

fib_iter(2, A, B, Result) :-
    Result #= A + B.
fib_iter(N, A, B, Result) :-
    N > 2,
    N1 is N - 1,
    C #= A + B,
    fib_iter(N1, B, C, Result).

% Matrix multiplication

matrix_multiply([], _, []).
matrix_multiply([Row|Matrix1], Matrix2, [ResultRow|ResultMatrix]) :-
    transpose(Matrix2, Matrix2T),
    row_multiply(Row, Matrix2T, ResultRow),
    matrix_multiply(Matrix1, Matrix2, ResultMatrix).

row_multiply(Row, [Col|Cols], [Sum|Rest]) :-
    dot_product(Row, Col, Sum),
    row_multiply(Row, Cols, Rest).
row_multiply(_, [], []).

dot_product([], [], 0).
dot_product([X|Xs], [Y|Ys], Sum) :-
    dot_product(Xs, Ys, Rest),
    Sum is X * Y + Rest.

% Generate random array with seed

generate_random_array(Size, Seed, Array) :-
    set_random(seed(Seed)),
    length(Array, Size),
    maplist(random(0, 10000), Array).

% Generate random matrix with seed

generate_random_matrix(Size, Seed, Matrix) :-
    set_random(seed(Seed)),
    length(Matrix, Size),
    maplist(generate_matrix_row(Size), Matrix).

generate_matrix_row(Size, Row) :-
    length(Row, Size),
    maplist(random(0, 100), Row).

% Main benchmark predicate

run_benchmarks(Result) :-
    % 1. Fibonacci Benchmark
    statistics(runtime, [StartFib|_]),
    fib(450, FibResult),
    statistics(runtime, [EndFib|_]),
    FibTime is (EndFib - StartFib),

    % 2. Sorting Benchmark

    % Note: Using SWI-Prolog built-in sort/2 instead of manual quicksort.
    % Rationale: The manual quicksort implementation caused stack overflow
    % for a 1,000,000-element list due to deep recursion and list copying.
    % sort/2 is optimized internally (typically a merge sort variation),
    % leveraging Prolog strengths and avoiding stack issues, though it
    % deviates from the battle usual requirement of manual quicksort.
    % This choice aligns with R and Julia use of built-in sort functions
    % for practicality in their respective benchmarks.

    generate_random_array(1000000, 42, Array),
    statistics(runtime, [StartSort|_]),
    sort(Array, _SortedArray),
    statistics(runtime, [EndSort|_]),
    SortTime is (EndSort - StartSort),

    % 3. Matrix Multiplication Benchmark
    generate_random_matrix(250, 42, Matrix1),
    generate_random_matrix(250, 42, Matrix2),
    statistics(runtime, [StartMatrix|_]),
    matrix_multiply(Matrix1, Matrix2, _ResultMatrix),
    statistics(runtime, [EndMatrix|_]),
    MatrixTime is (EndMatrix - StartMatrix),

    % Format results

    Result = [
        [fibonacci, [[n, 450], [result, FibResult], [time_ms, FibTime]]],
        [sorting, [[array_size, 1000000], [time_ms, SortTime]]],
        [matrix, [[matrix_size, 250], [time_ms, MatrixTime]]]
    ].

% Define result
res(Result) :- run_benchmarks(Result).



% Main predicate
main :-
    writeln("SWI-Prolog program starting..."),
    res(Res),
    writeln("Transforming result..."),
    transform_to_ring(Res, "plresult.txt"),
    writeln("Data written to file").

File: libraries/stzlib/base/test/modular/extercode/temp.jl

function transform_to_ring(data)
    function _transform(obj, depth=0)
        # Prevent excessive recursion
        if depth > 100
            return "TOO_DEEP"
        end
        
        # Handle nothing/missing values
        if obj === nothing
            return "NULL"
        end
        
        # Handle dictionaries
        if isa(obj, Dict)
            items = String[]
            for (key, value) in obj
                push!(items, "[\'$(key)\', $(_transform(value, depth + 1))]")
            end
            return "[" * join(items, ", ") * "]"
        end
        
        # Handle arrays
        if isa(obj, AbstractArray)
            return "[" * join([_transform(item, depth + 1) for item in obj], ", ") * "]"
        end
        
        # Handle strings
        if isa(obj, AbstractString)
            return "\'$(obj)\'"
        end
        
        # Handle numeric values
        if isa(obj, Number)
            str_val = string(obj)
            # Check for scientific notation
            if occursin(r"e|E", str_val)
                return "\'$(str_val)\'"
            end
            return str_val
        end
        
        # Handle boolean values
        if isa(obj, Bool)
            return obj ? "TRUE" : "FALSE"
        end
        
        # Default case: convert to string
        return "\'$(string(obj))\'"
    end
    
    return _transform(data)
end

# Main code
println("Julia script starting...")

# 1. Fibonacci function
function fib(n)
    if n <= 1
        return n
    end
    
    a, b = 0, 1
    for i in 2:n
        a, b = b, a + b
    end
    return b
end

# 2. Quicksort implementation
function quicksort!(arr, low, high)
    if low < high
        # Partition the array
        pivot = arr[high]
        i = low - 1
        
        for j in low:(high-1)
            if arr[j] < pivot
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
            end
        end
        
        arr[i+1], arr[high] = arr[high], arr[i+1]
        
        partition = i + 1
        
        # Recursively sort the sub-arrays
        quicksort!(arr, low, partition - 1)
        quicksort!(arr, partition + 1, high)
    end
    return arr
end

# Performance benchmark
using Random
using LinearAlgebra

results = []

# 1. Fibonacci sequence benchmark
n = 450
start_time = time()
result = fib(n)
end_time = time()
fib_time = (end_time - start_time) * 1000  # Convert to milliseconds

push!(results, ["fibonacci", [
    ["n", n],
    ["result", result],
    ["time_ms", fib_time]
]])

# 2. Sorting benchmark
start_time = time()
array_size = 1000000  # One million elements
array = zeros(Int, array_size)

# Fill with random numbers
Random.seed!(42)  # Fixed seed for reproducibility
for i in 1:array_size
    array[i] = rand(0:9999)
end

# Sort the array (using Julia built-in sort! for performance on large arrays)
sort!(array)
end_time = time()
sort_time = (end_time - start_time) * 1000  # Convert to milliseconds

push!(results, ["sorting", [
    ["array_size", array_size],
    ["time_ms", sort_time]
]])

# 3. Matrix multiplication benchmark
start_time = time()
matrix_size = 250

# Initialize matrices
Random.seed!(42)
matrix1 = rand(0:99, matrix_size, matrix_size)
matrix2 = rand(0:99, matrix_size, matrix_size)

# Matrix multiplication (using Julia built-in multiplication)
result_matrix = matrix1 * matrix2

end_time = time()
matrix_time = (end_time - start_time) * 1000  # Convert to milliseconds

push!(results, ["matrix", [
    ["matrix_size", matrix_size],
    ["time_ms", matrix_time]
]])

# Return results
res = Dict(
    "fibonacci" => Dict(
        "n" => n,
        "result" => result,
        "time_ms" => fib_time
    ),
    "sorting" => Dict(
        "array_size" => array_size,
        "time_ms" => sort_time
    ),
    "matrix" => Dict(
        "matrix_size" => matrix_size,
        "time_ms" => matrix_time
    )
)
transformed = transform_to_ring(res)
println("Data before transformation: ", res)
println("Data after transformation: ", transformed)
open("jlresult.txt", "w") do f
    write(f, transformed)
end
println("Data written to file")
